src/libmongoc/src/mongoc/mongoc-ts-pool.c

```diff
@@ -0,0 +1,263 @@
+#include "mongoc-ts-pool-private.h"
+#include "common-thread-private.h"
+
+#include "bson/bson.h"
+
+#if defined(_MSC_VER)
+typedef double _max_align_type;
+#elif defined(__APPLE__)
+typedef long double _max_align_type;
+#else
+/* Based on the GCC max_align_t definition */
+typedef struct _max_align_type {
+   long long maxalign_1 __attribute__ ((__aligned__ (__alignof__(long long))));
+   long double maxalign_2
+      __attribute__ ((__aligned__ (__alignof__(long double))));
+} _max_align_type;
+#endif
+
+typedef struct pool_node {
+   struct pool_node *next;
+   mongoc_ts_pool *owner_pool;
+
+   _max_align_type data[0];
```

Comment:
Rather than emulating flexible array members, wouldn't simply using them be a viable alternative, given the C driver is already making use of C99 features?
```suggestion
   _max_align_type data[];
```
GCC's documentation for [Arrays of Length Zero](https://gcc.gnu.org/onlinedocs/gcc/Zero-Length.html) also recommends using C99 flexible array members instead of the extension:

> The preferred mechanism to declare variable-length types like `struct line` above is the ISO C99 _flexible array member_, with slightly different syntax and semantics: [...]

This also applies to `mongoc_ts_pool::outstanding_items`.

Reply:
In this case, using an FAM is entirely reasonable, given that we already have some C99 peppered throughout.

For `outstanding_items`, I'd be sad to change it, because having that member be present/absent based on a `1/0` enum that can also be used in `if ()` conditions is so elegant (when compared to using an `#ifdef`) 😢. Alternatively, just bumping the size of of the `malloc()` based on `AUDIT_POOL_ENABLED` isn't a huge burden either 🙂. Or even just foregoing that trickery and letting the member be unused unless `AUDIT_POOL_ENABLED != 0`. (This is interesting enough that I might do a write-up on it.)

Reply:
I just removed the cleverness with `outstanding_items`. It's only a single integer, so not worth trying to be tricky.